pkg/webui/console/views/device-add/index.js

```diff
@@ -0,0 +1,383 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import React, { Component } from 'react'
+import { Container, Col, Row } from 'react-grid-system'
+import bind from 'autobind-decorator'
+import { push } from 'connected-react-router'
+import { connect } from 'react-redux'
+
+import Form from '../../../components/form'
+import Field from '../../../components/field'
+import Button from '../../../components/button'
+import Breadcrumb from '../../../components/breadcrumbs/breadcrumb'
+import { withBreadcrumb } from '../../../components/breadcrumbs/context'
+import FieldGroup from '../../../components/field/group'
+import Message from '../../../lib/components/message'
+import IntlHelmet from '../../../lib/components/intl-helmet'
+import SubmitBar from '../../../components/submit-bar'
+import toast from '../../../components/toast'
+
+import api from '../../api'
+import sharedMessages from '../../../lib/shared-messages'
+import { get as cacheGet, set as cacheSet } from '../../lib/cache'
+import m from './messages'
+import validationSchema from './validation-schema'
+import style from './device-add.styl'
+
+@withBreadcrumb('devices.add', function (props) {
+  const { appId } = props.match.params
+  return (
+    <Breadcrumb
+      path={`/console/applications/${appId}/devices/add`}
+      icon="add"
+      content={sharedMessages.add}
+    />
+  )
+})
+@connect()
+@bind
+export default class DeviceAdd extends Component {
+  state = {
+    error: '',
+    otaa: true,
+    resets_join_nonces: false,
+    resets_f_cnt: false,
+    frequencyPlanOptions: [],
+  }
+
+  async componentDidMount () {
+    let frequencyPlans = cacheGet('frequency_plans')
+    try {
+      if (!frequencyPlans) {
+        frequencyPlans = (await api.configuration.listNsFrequencyPlans()).frequency_plans
+        cacheSet('frequency_plans', frequencyPlans)
```

Comment:
why not storing freq plans in redux? do we really need this?

Reply:
Moreover, we should try to split the presentation and data fetching as much as possible. Note, that the forms are different.

Reply:
This is to persist the data between reloads. @htdvisser notified me that it makes sense to cache the config since the config data will not change frequently. I will move the retrieval logic into redux.